src/hotspot/share/gc/g1/g1ServiceThread.hpp

```diff
@@ -28,43 +28,83 @@
 #include "gc/shared/concurrentGCThread.hpp"
 #include "runtime/mutex.hpp"
 
+class G1ServiceTask : public CHeapObj<mtGC> {
+  // The next time this task should be executed.
+  double _time;
+  // Name of the task.
+  const char* _name;
+  // Next task in the task list.
+  G1ServiceTask* _next;
+
+public:
+  G1ServiceTask(const char* name);
+  const char* name();
+
+  void set_time(double time);
+  double time();
+
+  void set_next(G1ServiceTask* next);
+  G1ServiceTask* next();
+
+  // Do the actual work for the task.
+  virtual void execute() = 0;
+  // Timeout to the next invocation. A negative value can be used
+  // to stop the task from being rescheduled and run again.
+  virtual int64_t timeout_ms() = 0;
+};
+
+class G1SentinelTask : public G1ServiceTask {
+public:
+  G1SentinelTask();
+  virtual void execute();
+  virtual int64_t timeout_ms();
+};
+
+class G1ServiceTaskList {
+  // The sentinel task is the entry point of this ordered circular list holding
+  // the service tasks. The list is ordered by the time the tasks are scheduled
+  // to run and the sentinel task has the time set to DBL_MAX. This guarantees
+  // that any new task will be added just before the sentinel at the latest.
+  G1SentinelTask _sentinel;
+
+  // Verify that the list is ordered.
+  void verify_task_list() NOT_DEBUG_RETURN;
+public:
+  G1ServiceTaskList();
+  G1ServiceTask* pop();
+  G1ServiceTask* peek();
+  void add_ordered(G1ServiceTask* task);
+  bool is_empty();
```

Comment:
I think it is better we call it  a priority queue instead of a list.

Reply:
Sound good, I'll change it to `G1ServiceTaskQueue` and mention it's a priority queue in the comment.